Add support for other id3 tools

thaffenden opened the pull request:
I had issues using this, as the version of `id3` I had didn't like the `-q` flag, so thought I'd update it to use `id3tool` as mentioned in #4 .
So I got that working although it turns out that didn't actually fix what I needed it to, as `id3tool` only writes v1 tags (probbbbably should have checked that first :facepalm: , but oh well).

So I'm also going to update it to handle using `id3v2` in the same manner (which **will** fix my use case), but figured it was worth keeping the `id3tool` bits in there as it fixes an issue someone else raised.

thaffenden commented:
Updated to add support for `id3v2` as well.
I've left the format in the vim buffer the same for all of them so editing is a consisten experience, but I found getting the tag values by the index (as both `id3` and `id3tool` do in their read methods) to be unreliable for `id3v2` as the index would change for different files based on the tags set on the file, so I wrote a helper to grab them by their id3 v2 spec values (which can be listed with `id3v2 --list-frames`.

If this all looks good I'll update the README as well.

AndrewRadev commented:
Thank you for your work! The `id3` tool occasionally even errors out for me for some filenames, so it's probably time for me to try something new myself.

I usually avoid using lambda functions in plugins (the `trim()` function is from 8.0.x as well, surprisingly enough), but it's actually been quite enough of a while since 7.4 that most distros ought to have something more up-to-date. In the worst case, if somebody opens an issue, I can take care of the mechanical work to achieve compatibility, it shouldn't be too big of a problem.

As long as this works for you with your id3v2 setup, I'd appreciate you updating the README, too, and then I'll be happy to merge :+1:

thaffenden commented:
I've been updating a bunch of files with it for the past week, found a coupld of issues which I have since fixed (it erroring with multiline comment strings, and didn't support the three character ID3v2.2 tags).
I've updated it so it will now get the data for any old ID3v2.2 (3 character) tags, but when you write the buffer it will save them as ID3v2.3 (4 character) tags, which I figured should be fine as it's keeping the data just updating the tag version used to a more recent one.

I'm also currently setting the value of `TPE2` (band/orchestra/accompaniment) to the same as the value for `TPE1` (lead performer(s)/soloist(s) - used as the `Artist` tag), as I had some files where they were set differently but in reality I needed them as the same value. I guess other people might not want that, particularly with clasical music. So I could take that bit out, or expose it as a new value in the buffer (so the consistency of the editing UI would start to deviate slightly)? But for my use case I think I'd always be setting those to the same values.

I've updated the README with some more info, but if you can think of anything else it needs etc just let me know!
If I run into any more issues after this is merged, or anyone else reports any problems around it I'm happy to raise PRs to fix.

AndrewRadev commented:
> when you write the buffer it will save them as ID3v2.3 (4 character) tags, which I figured should be fine as it's keeping the data just updating the tag version used to a more recent one.

Yeah, this is fine, I've even had situations, where I wanted this, having both kinds of tags and wanting to unify them, when written with Audacity (or at least I think that's what was happening. Reattaching the cover image with my [id3-image](https://github.com/AndrewRadev/id3-image) seemed to fix it).

> I'm also currently setting the value of TPE2 (band/orchestra/accompaniment) to the same as the value for TPE1 (lead performer(s)/soloist(s) - used as the Artist tag), as I had some files where they were set differently but in reality I needed them as the same value. I guess other people might not want that, particularly with clasical music. So I could take that bit out, or expose it as a new value in the buffer (so the consistency of the editing UI would start to deviate slightly)?

Hmm, if a write operation can delete some extra data, that doesn't seem ideal. How do you feel about adding a setting for that, something like:

``` vim
if !exists('g:id3_merge_tpe1_and_tpe2')
  let g:id3_merge_tpe1_and_tpe2 = 0
endif
```

In the `plugin/id3.vim` file, above the autocommands? It could default to 0 and be set to 1 if somebody like you wants to do this.

I'm also not opposed to a different interface for the different backends, since each callback currently sets up its own buffer-writing and parsing anyway, so it wouldn't be a big problem in terms of code reuse. Maybe apart from the human-readable names, we could have stuff like `(TPE1)` in brackets or something. But I'd say that's not important for this PR in particular.

AndrewRadev commented:
I just realized I still haven't merged this. Could you comment on this note?

> Hmm, if a write operation can delete some extra data, that doesn't seem ideal. How do you feel about adding a setting for that, something like:
> 
>     if !exists('g:id3_merge_tpe1_and_tpe2')
>       let g:id3_merge_tpe1_and_tpe2 = 0
>     endif
>
> In the plugin/id3.vim file, above the autocommands? It could default to 0 and be set to 1 if somebody like you wants to do this.

thaffenden commented:
@AndrewRadev sorry, complete dropped the ball on this one and forgot about it until I got an email about the reference in #6.

I've rebased the PR to get it up to date, and removed the line that set the `TPE2` field to the same value as `TPE1`. Sanity tested on a couple of files with `id3v2` and all seems fine.

I thought I'd remove rather than adding it as a config setting as it is a bit of a strange use case related to specifically what I was trying to do with some existing media I had rather than it being a common thing. Easy enough to add in a new PR if anyone requests though!

AndrewRadev commented:
No worries about taking a while, this kind of thing happens with open source occasionally :D. Thanks for your changes, I'll merge them for now and see if I can add the ordered list of configs. For the TPE, as we discussed in #6, maybe it would make sense to change the UI to whatever the tool provides, but I'll start from what we've got here and experiment :+1:
